src/webgpu/api/validation/query_set/create.spec.ts

```diff
@@ -0,0 +1,141 @@
+export const description = `
+Tests for validation in createQuerySet.
+`;
+
+import { poptions } from '../../../../common/framework/params_builder.js';
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { ValidationTest } from '../validation_test.js';
+
+export const g = makeTestGroup(ValidationTest);
+
+const kMaxQueryCount = 8192;
+
+// Test with these query counts.
+const kQueryCounts: number[] = [
+  // valid
+  0,
+  kMaxQueryCount,
+  // invalid
+  kMaxQueryCount + 1,
+];
+
+g.test('occlusion')
+  .desc(
+    `
+Tests that create query set with the type of occlusion.
+- control cases:
+  - count <= kMaxQueryCount
+  - pipelineStatistics is undefined or empty
+- invalid cases:
+  - count > kMaxQueryCount
+  - pipelineStatistics is set with pipeline statistic names
+  `
+  )
+  .params([
+    ...poptions('count', kQueryCounts),
+    ...poptions('pipelineStatistics', [
+      // valid
+      undefined,
+      [],
+      // invalid
+      ['clipper-invocations'],
+    ]),
+  ])
+  .fn(t => {
+    const { count = 1, pipelineStatistics } = t.params as {
+      count?: number;
+      pipelineStatistics?: GPUPipelineStatisticName[];
+    };
+
+    t.expectValidationError(() => {
+      t.device.createQuerySet({ type: 'occlusion', count, pipelineStatistics });
+    }, count > kMaxQueryCount || (pipelineStatistics !== undefined && pipelineStatistics.length > 0));
+  });
+
+g.test('pipeline_statistics')
+  .desc(
+    `
+Tests that create query set with the type of pipeline-statistics.
+- control cases:
+  - count <= kMaxQueryCount
+  - pipelineStatistics is set with pipeline statistic names
+- invalid cases:
+  - count > kMaxQueryCount
+  - pipelineStatistics is undefined, empty or duplicated
+  `
+  )
+  .params([
+    ...poptions('count', kQueryCounts),
+    ...poptions('pipelineStatistics', [
+      // valid
+      ['clipper-invocations'],
+      // invalid
+      undefined,
+      [],
+      ['clipper-invocations', 'clipper-invocations'],
+    ]),
+  ])
+  .fn(async t => {
+    await t.selectDeviceOrSkipTestCase({
+      extensions: ['pipeline-statistics-query'],
+    });
+
+    const { count = 1, pipelineStatistics = ['vertex-shader-invocations'] } = t.params as {
```

Comment:
I'm not sure I understand, why do we need the defaulting to vertex-shader-invocations to then detect that case and go back to setting undefined?

Reply:
+1, I don't think any of that is necessary.

Reply:
We use `...poptions` for setting the parameters count and pipelineStatistics. When testing count, pipelineStatistics is undefined, and vice versa, so we need to set a valid default value for them. But for pipeline statistics query, undefined is also a checkpoint, which overwrite by the default value, so I set different value from the parameter, and detect it should be checked as undefined. 

Kai's suggestion may avoid such problem.

Reply:
Ah I see I didn't really think about the fact that this is [...poptions] instead of .combine(poptions).

We should definitely rephrase this so that this complexity isn't needed. Probably by simply splitting this test into two - one for .count and one for .pipelineStatistics - as I think it would be simpler.